internal/controller/pkg/resolver/reconciler.go

```diff
@@ -298,14 +273,63 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	// no packagePullSecrets are set. Settings can be modified manually
 	// after dependency creation to address this.
 	pack.SetName(xpkg.ToDNSLabel(ref.Context().RepositoryStr()))
-	pack.SetSource(fmt.Sprintf(packageTagFmt, ref.String(), addVer))
+
+	format := packageTagFmt
+	if strings.Contains(addVer, "sha256") {
+		format = packageDigestFmt
+	}
+
+	pack.SetSource(fmt.Sprintf(format, ref.String(), addVer))
 
 	// NOTE(hasheddan): consider making the lock the controller of packages
 	// it creates.
-	if err := r.client.Create(ctx, pack); err != nil {
+	if err := r.client.Create(ctx, pack); !kerrors.IsAlreadyExists(err) {
 		log.Debug(errCreateDependency, "error", err)
 		return reconcile.Result{}, errors.Wrap(err, errCreateDependency)
 	}
 
 	return reconcile.Result{Requeue: false}, nil
 }
+
+func (r *Reconciler) findDependencyVersion(ctx context.Context, dep *v1beta1.Dependency, log logging.Logger, ref name.Reference) (string, error) {
+	var addVer string
+
+	if strings.HasPrefix(dep.Constraints, "sha256:") {
+		log.Debug("package is pinned to a specific digest, skipping resolution")
+		return dep.Constraints, nil
```

Comment:
nit: it would be good to update the [comment](https://github.com/crossplane/crossplane/blob/edbc23e9ec3a676726af64bb77ae61b2992c97d8/apis/pkg/v1beta1/lock.go#L97-L99) on Constraints field to indicate that it can be either a semver range or a digest.

Reply:
added